### PR TITLE
expander optimization

### PR DESCRIPTION
**Pull request by SimaTian**
# changes:
 1. The AddArgumentFromSlices is an optimization that bypasses the need for argumentBuilder which has noticeable overhead.
 I use indices to directly calculate the argument instead.

 2. The lines 2126 - 2142
  The `expressionCapture.Captures?.Any ` iteration was visible in the profiler.
  I pulled both branches into one loop since it looked slightly better. Performance improvement is speculative at best, can be 
  removed if we decide that it is too risky or something.

3. Finally the `src/StringTools/WeakStringCache.Concurrent.cs`
   the .Count call on the ConcurrentDictionary is locking everything which is not great since we do it quite often. my change aims to introduce separate count for that to limit the number of locks we take and to unblock the dictionary.

All together when profiling on my machine, after killing the outliers (MSBuild sometimes has a random-ish spikes in build time that I decided to skip) my aggregated output was ~this:

1. 23 cold runs for our current Main branch, with `-maxcpucount:10 -tl:true` and average time of 62.103s
2. 13 cold runs for this branch with same arguments, average time was 61.31s
Though these are speculative at best, I've yet to find a reasonable and stable enough way of checking the impact.

While using VS profiler for running my changes on a single-process basis, the improvement looked to be somewhere in the vicinity of 0.2-03%

### CODE REVIEW

**Comment by maridematte** [comment]
Thank you for tackling performance, I just have a few comments before getting to the specifics of the code.

1. There are a lot of commented out code that doesn't seem that will be used in the future, please clean it up.
2. For the PR description, it is very confusing. Please add the reason that these changes need to be made, and separate what you changed and the performance impact into two separate sections.

**Comment by SimaTian** [comment]
> Thank you for tackling performance, I just have a few comments before getting to the specifics of the code.
> 
> 1. There are a lot of commented out code that doesn't seem that will be used in the future, please clean it up.
> 2. For the PR description, it is very confusing. Please add the reason that these changes need to be made, and separate what you changed and the performance impact into two separate sections.

 1. I originally intended to keep the old code around for comparison purposes. That being said I forget that Git will make it visible anyways so I will remedy that soon-ish.
 2. I will try.

File: src/StringTools/WeakStringCache.Concurrent.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ namespace Microsoft.NET.StringTools
     internal sealed partial class WeakStringCache : IDisposable
     {
         private readonly ConcurrentDictionary<int, StringWeakHandle> _stringsByHashCode;
+        private int _count;
```
Comment (JanKrivanek) [inline src/StringTools/WeakStringCache.Concurrent.cs:17]: This feels can get out of sync after the `Scavenge` call (as judging by name, the goal is to remove items, while we only increment the count)

File: src/Build/Evaluation/Expander.cs [hunk 2]
```diff
@@ -830,18 +875,14 @@ private static string[] ExtractFunctionArguments(IElementLocation elementLocatio
 
             List<string> arguments = new List<string>();
 
-            using SpanBasedStringBuilder argumentBuilder = Strings.GetSpanBasedStringBuilder();
-            int? argumentStartIndex = null;
+            int abStart = -1;
```
Comment (JanKrivanek) [inline src/Build/Evaluation/Expander.cs:878]: Can this be named more expressively?

File: src/Build/Evaluation/Expander.cs [hunk 3]
```diff
@@ -775,47 +772,95 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
+                firstSliceIdx++;
+                if (firstSliceIdx > slices[firstSlice].Item2 && ++firstSlice < lastSlice)
+                {
+                    firstSliceIdx = slices[firstSlice].Item1;
+                }
             }
-            else
+
+            // Trim from the end.
+            // Bit of extra logic to avoid trimming whitespace-only string one time too many.
```
Comment (JanKrivanek) [inline src/Build/Evaluation/Expander.cs:804]: What does this mean exactly? :)

File: src/Build/Evaluation/Expander.cs [hunk 4]
```diff
@@ -775,47 +772,95 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
+
+            bool removedQuotes = false;
+
+            if ((arg[firstSliceIdx] == '\'' && arg[lastSliceIdx - 1] == '\'') ||
+                (arg[firstSliceIdx] == '`' && arg[lastSliceIdx - 1] == '`') ||
+                (arg[firstSliceIdx] == '`' && arg[lastSliceIdx - 1] == '`') ||
+                (arg[firstSliceIdx] == '"' && arg[lastSliceIdx - 1] == '"'))
+            {
+                ++firstSliceIdx;
+                --lastSliceIdx;
```
Comment (JanKrivanek) [inline src/Build/Evaluation/Expander.cs:828]: This deserves comment (e.g. 'skiping the initial and trialing enclosing chars')

**Review by JanKrivanek — COMMENTED** [review]
Thanks for digging into this!!

The main reason why I do not want to approve now - This is nontrivial change in core logic and the PR has no mention of test coverage. If it's already covered by pre-existing tests, please call those out. If not - I'd prefer a tailored test(s) to be added first (can be same PR, but idealy a separate commit, that shows the test were passing on previous version).

File: src/StringTools/WeakStringCache.Concurrent.cs (hunk 1 above)
Comment (SimaTian) [reply src/StringTools/WeakStringCache.Concurrent.cs:17]: The scavenge call is resetting this counter after it's done:
```
_scavengeThreshold = _stringsByHashCode.Count * 2;
                      _count = _stringsByHashCode.Count;
```
There might be a slight inaccuracy, however I would argue that it's fine due to the fact that we're only using it here and only as a threshold to occasionaly reduce the size of the collection.

File: src/Build/Evaluation/Expander.cs (hunk 4 above)
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:828]: No problem, will do.
This function is nearly exact match of the previous version, just rewritten to use indices - so I mostly just copied comments over from that one. Probably should've thought about it a bit more.

File: src/Build/Evaluation/Expander.cs (hunk 3 above)
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:804]: When the string is whitespace only, this would first trim from the start and trim everything, then trim one more character from the back due to the way indices are set up here. (If I trimmed naively)
This is a bit of extra logic to prevent that.
Don't ask me how I found out.

**Comment by SimaTian** [comment]
Regarding the test coverage:
We have 348 expander tests covering this area in general, see Expander_tests.cs
If I were to call out one test in particular that made my life somewhat uncomfortable, it would be `Medley` - a bunch of random error checks and edge cases.
Every test there does some expansion somewhere.
Furthermore, I was able to build the Orchard even while some tests were failing - I don't think I can do much better test wise, unless we want to spend a lot more time on this.

File: src/StringTools/WeakStringCache.Concurrent.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/StringTools/WeakStringCache.Concurrent.cs:17]: I have for some reason missed that part :-O
Looks good like this!

**Comment by JanKrivanek** [comment]
All sounds perfect!

Can you run an exp insertion, request perf runs (DDRITs and Speedometer) a post results here?

**Review by JanKrivanek — APPROVED** [review]
Thank you!

File: src/Build/Evaluation/Expander.cs [hunk 5]
```diff
@@ -775,47 +772,99 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
         /// appropriately.
         /// </summary>
-        private static void AddArgument(List<string> arguments, SpanBasedStringBuilder argumentBuilder)
+        private static void AddArgumentFromSlices(List<string> arguments, List<Tuple<int, int>> slices, string arg)
         {
-            // we reached the end of an argument, add the builder's final result
-            // to our arguments.
-            argumentBuilder.Trim();
-            string argValue = argumentBuilder.ToString();
+            // This shouldn't happen
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:780]: please elaborate on "why this shouldn't happen"

File: src/Build/Evaluation/Expander.cs [hunk 6]
```diff
@@ -775,47 +772,99 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
+                    firstSlice++;
+                    firstSliceIdx = slices[firstSlice].Item1;
+                }
             }
-            else
+
+            // Trim from the end.
+            // There is some extra logic here to avoid edge case where we would trim a whitespace only character with
+            // one slice from the start, and then once more from the end, resulting in invalid indices.
+            while (((firstSlice < lastSlice) || (firstSlice == lastSlice && firstSliceIdx < lastSliceIdx)) && Char.IsWhiteSpace(arg, lastSliceIdx - 1))
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:807]: it looks like `(firstSlice <= lastSlice && Char.IsWhiteSpace(arg, firstSliceIdx)` can be moved to the separate method with descriptive name and reused in the while loop here and on the line 794

File: src/Build/Evaluation/Expander.cs [hunk 7]
```diff
@@ -2012,6 +2061,23 @@ internal static bool ExpandExpressionCapture<S>(
                 // If there are no items of the given type, then bail out early
                 if (itemsOfType.Count == 0)
                 {
+                    if (expressionCapture.Captures?.Any(capture =>
+                        {
+                            if (string.Equals(capture.FunctionName, "Count", StringComparison.OrdinalIgnoreCase))
+                            {
+                                return true;
+                            }
+                            if (string.Equals(capture.FunctionName, "AnyHaveMetadataValue", StringComparison.OrdinalIgnoreCase))
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:2070]: can it be just `return string.Equals(capture.FunctionName, "AnyHaveMetadataValue", StringComparison.OrdinalIgnoreCase)`?

File: src/Build/Evaluation/Expander.cs [hunk 8]
```diff
@@ -775,47 +772,99 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
+                }
+                if (slices[lastSlice].Item1 > lastSliceIdx && firstSlice < --lastSlice)
+                {
+                    lastSliceIdx = slices[lastSlice].Item2;
                 }
             }
+
+            string argValue = "";
+
+            while (firstSlice < lastSlice)
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:848]: could you please comment on these iterations so it's clear what  is handled here?

File: src/Build/Evaluation/Expander.cs [hunk 9]
```diff
@@ -2012,6 +2061,23 @@ internal static bool ExpandExpressionCapture<S>(
...
+                            if (string.Equals(capture.FunctionName, "Count", StringComparison.OrdinalIgnoreCase))
+                            {
+                                return true;
+                            }
+                            if (string.Equals(capture.FunctionName, "AnyHaveMetadataValue", StringComparison.OrdinalIgnoreCase))
+                            {
+                                return true;
+                            }
+                            return false;
+                        }) != true)
```
Comment (JanKrivanek) [inline src/Build/Evaluation/Expander.cs:2075]: Is it this?

```suggestion
                    if (expressionCapture.Captures?.Any(capture =>
                        !string.Equals(capture.FunctionName, "Count", StringComparison.OrdinalIgnoreCase) ||
                        !string.Equals(capture.FunctionName, "AnyHaveMetadataValue", StringComparison.OrdinalIgnoreCase)
                        )
```

File: src/Build/Evaluation/Expander.cs [hunk 10]
```diff
@@ -775,47 +772,99 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
+            {
+                arguments.Add("");
+                return;
+            }
+
+            bool removedQuotes = false;
+
+            // If the argument is in quotes, we want to remove those
+            if ((arg[firstSliceIdx] == '\'' && arg[lastSliceIdx - 1] == '\'') ||
+                (arg[firstSliceIdx] == '`' && arg[lastSliceIdx - 1] == '`') ||
```
[the remark this answers is not in the document]
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:827]: The comparison didn't use them (I'm not 100% sure why, but I kept it that way). Since I'm working with indices, I didn't need them for .Trim() calls anymore so they became unused and the ./build.cmd started complaining.

File: src/Build/Evaluation/Expander.cs [hunk 11]
```diff
@@ -775,47 +772,99 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
+                (arg[firstSliceIdx] == '`' && arg[lastSliceIdx - 1] == '`') ||
+                (arg[firstSliceIdx] == '"' && arg[lastSliceIdx - 1] == '"'))
+            {
+                ++firstSliceIdx;
+                --lastSliceIdx;
+                removedQuotes = true;
+
+                // Check yet again if we're still in the correct slice boundaries, this could've changed if we've trimmed.
+                if (firstSliceIdx > slices[firstSlice].Item2 && ++firstSlice < lastSlice)
```
[the remark this answers is not in the document]
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:836]: Fair point. I think I removed it in another place but this one slipped through the crack.

**Comment by YuliiaKovalova** [comment]
@SimaTian , I am slightly concerned about code readability factor.
It's great that applying these changes brings  perf improvements , but it's quite difficult to understand a new logic in the amended parts.

File: src/Build/Evaluation/Expander.cs [hunk 12]
```diff
@@ -830,18 +879,14 @@ private static string[] ExtractFunctionArguments(IElementLocation elementLocatio
 
             List<string> arguments = new List<string>();
 
-            using SpanBasedStringBuilder argumentBuilder = Strings.GetSpanBasedStringBuilder();
-            int? argumentStartIndex = null;
+            int argumentStartIndex = -1;
+            List<Tuple<int, int>> slices = new List<Tuple<int, int>>();
```
[the remark this answers is not in the document]
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:883]: There is a ReadOnlySpan<char> that would achieve this already - found that one after I've had this one done, which is admittedly a poor excuse.
We have actually a whole ticket about replacing strings with ReadOnlySpan<char> wherever possible - it's the next one I will be working on. Unfortunately there are some compatibility issues due to TaskHost being stuck with old net35.

File: src/Build/Evaluation/Expander.cs [hunk 13]
```diff
@@ -878,32 +923,36 @@ void FlushCurrentArgumentToArgumentBuilder(int argumentEndIndex)
...
+                    AddArgumentFromSlices(arguments, slices, argumentsString);
 
                     // Clear out the argument builder ready for the next argument
-                    argumentBuilder.Clear();
+                    slices.Clear();
                 }
                 else
                 {
-                    argumentStartIndex ??= n;
+                    // argumentStartIndex ??= n;
```
[the remark this answers is not in the document]
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:942]: fair point, I've removed most of these, this one slipped through.

File: src/Build/Evaluation/Expander.cs [hunk 14]
```diff
@@ -775,47 +772,99 @@ private static int ScanForClosingQuote(char quoteChar, string expression, int in
...
+                {
+                    lastSliceIdx = slices[lastSlice].Item2;
                 }
             }
+
+            string argValue = "";
+
+            while (firstSlice < lastSlice)
+            {
+                argValue += arg.Substring(firstSliceIdx, slices[firstSlice].Item2 - firstSliceIdx);
```
[the remark this answers is not in the document]
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:850]: As for the perf improvement, the first two lines of the previous version function were doing this:
```
  argumentBuilder.Trim();
  string argValue = argumentBuilder.ToString();
``` 
and then doing everything with String, throwing away any and all advantage the spanbased stringbuilder might've had.
So there is an option to kill most of my changes and replace the string with SpanBasedChar - it could achieve similar if not identical results. The main cost would be probably further profiling since it's rather large replacement once again.

File: src/Build/Evaluation/Expander.cs (hunk 5 above)
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:780]: There should always be at least one slice. If something goes here then something is wrong.
Maybe I should've been more explicit and throw instead or remove the check alltogether since I don't think it can happen.
It was mostly me covering my bases when starting to fiddle around with indices.

File: src/Build/Evaluation/Expander.cs (hunk 6 above)
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:807]: This checks checks against lastSliceIndx and also checks to prevent duplicate whitespace removal (the second part after the ||)
It was similar-ish before, then I run into the "all whitespace" edge case and crashed. 
that is what the `firstSlice == lastSlice && firstSliceIdx < lastSliceIdx` is for.
So the function would have to be something like
`IsEdgeWhitespace(arg, firstSlice, lastSlice, idxToCheck)` and then some extra logic for this case.
Is that more readable? I would like to say no, but that could be just my laziness/lack of feeling for cases such as this.

File: src/Build/Evaluation/Expander.cs (hunk 9 above)
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:2075]: My brain is kind of fried at the moment, so I pasted this into .netfiddle:
```
using System;
using System.Collections.Generic;
using System.Linq;
					
public class Program
{
	public static void Main()
	{
		 var x = new List<string>();
		 x.Add("Count");
		 if (x.Any(capture =>
                        {
                            if (string.Equals(capture, "Count", StringComparison.OrdinalIgnoreCase))
                            {
                                return true;
                            }
                            if (string.Equals(capture, "AnyHaveMetadataValue", StringComparison.OrdinalIgnoreCase))
                            {
                                return true;
                            }
                            return false;
                        }) != true)
		 {
		 	Console.WriteLine("hit1");
		 }
	    if (x.Any(capture =>
												 !string.Equals(capture, "Count", StringComparison.OrdinalIgnoreCase) ||
												 !string.Equals(capture, "AnyHaveMetadataValue", StringComparison.OrdinalIgnoreCase)))
			{
			Console.WriteLine("hit2");
			}
	}
}
```
These aren't equivalent - empty collection hits only the first case, collection with "Count" hits only the second one.
I'll look at it closer when refactoring, some simplification along these lines is definitely possible, but I found myself repeatedly getting lost in the logic so in the end I simplified at the cost of verbosity.

File: src/Build/Evaluation/Expander.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:850]: Bit more googling and having a static StringBuilder and reusing the instance should have the best from the both worlds - reasonable performance for small concatenation counts while avoiding the danger of allocation.

File: src/Build/Evaluation/Expander.cs (hunk 9 above)
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs:2075]: Yuliia suggested a similar change with a similar simplification. In the end I went with that one.

**Comment by maridematte** [comment]
I agree with @YuliiaKovalova. The code readability is not great here, it took me sometime to understand what is going on, and this concerns me if we need to debug it. My current opinion is that the performance improvement on this one is not worth the code readability trade-off.

**Comment by SimaTian** [comment]
I will split out the two smaller non-problematic improvements to a separate PR and kill this one.
